### PR TITLE
The lamps support

### PR DESCRIPTION
**Pull request by parar020100**
I was keeping this on hold since I wanted to add the dynamic lamps / changing colors functionality
However, I think it should be merged as is (since the support for lamps IS ready)
The light states and stuff will probably come later

**How to test:** 2010 addon, `lamps-lighting` branch

**Description:**
adds the support for the use of lamps (projected lights) in the interiors

**Adding lamps to metadata:**
```lua
	Lamps = {
		-- under the console
		{
			color = Color(0, 255, 255),
			texture = "effects/flashlight/soft",
			fov = 169.09091186523,
			distance = 661.45452880859,
			brightness = 1.5909091234207,
			pos = Vector(0, 0, -50),
			ang = Angle(90, -180, 180),
		},

		-- right wall front
		{
			color = Color(255, 125, 35),
			texture = "effects/flashlight/soft",
			fov = 156.36363220215,
			distance = 661.45452880859,
			brightness = 0.75,
			pos = Vector(6.7666015625, 42.2294921875, 164.078125),
			ang = Angle(28.094324111938, 94.355094909668, -131.17744445801),
		},
	}
```

A convar `tardis2_debug_lamps` makes the lamps appear physically (they are invisible by default and can not be touched):

gm_freespace_130105

An extension creator can move the lamps around, change the position / color / FOV / ...
Then, after the lamp is clicked, it will generate the lua code for adding that lamp in that exact state:

image

**The possibilities** (examples):

gm_freespace_130098

gm_freespace_130099

gm_freespace_130100

gm_freespace_130102

gm_freespace_130104

### CODE REVIEW

File: lua/entities/gmod_tardis_interior/modules/sh_interior_lights.lua [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+                    render.SetMaterial(self.roundthingmat)
+                    render.DrawSprite(pos, size, size, Color(255,153,0, vis))
+                end
+            end
+        end
+    end)
+
+else -- SERVER
+    
+    if SERVER then
```
Comment (AmyJeanes) [inline lua/entities/gmod_tardis_interior/modules/sh_interior_lights.lua]: Redundant `if SERVER then` check

File: lua/entities/gmod_tardis_interior/modules/sh_interior_lights.lua [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        ENT:AddHook("Initialize", "lamps", function(self)
+            local lamps = self.metadata.Interior.Lamps
+            if not lamps then return end
+
+            self.lamps = {}
+
+            for k,v in pairs(lamps) do
+                if v then
+                    local lamp = MakeLamp(nil, --self:GetCreator(),
```
Comment (AmyJeanes) [inline lua/entities/gmod_tardis_interior/modules/sh_interior_lights.lua]: Why is self:GetCreator() commented out here?

**Review by AmyJeanes — CHANGES_REQUESTED** [review]
As mentioned on Discord this impacts performance too much and needs to be made clientside, if the feature can be added clientside without impacting performance much then we don't need an option for it

File: lua/entities/gmod_tardis_interior/modules/sh_interior_lights.lua (hunk 2 above)
Comment (parar020100) [reply lua/entities/gmod_tardis_interior/modules/sh_interior_lights.lua]: that's the "owner" parameter for the entity
I got rid of that due to the limit of entities per person
(if all TARDIS parts were owned by a player, he wouldn't be able to spawn many)

Comment (AmyJeanes) [reply lua/entities/gmod_tardis_interior/modules/sh_interior_lights.lua]: Interesting, okay can we remove the comment then?

Comment (parar020100) [reply lua/entities/gmod_tardis_interior/modules/sh_interior_lights.lua]: I'm removing the whole thing now

Comment (AmyJeanes) [reply lua/entities/gmod_tardis_interior/modules/sh_interior_lights.lua]: Will close this then as it's gonna need restructuring for the clientside stuff as we talked about

File: lua/tardis/sh_parts.lua [hunk 3]
```diff
@@ -28,7 +28,9 @@ function TARDIS.DrawOverride(self,override)
             if self.PreDraw then self:PreDraw() end
             if self.UseTransparencyFix and (not override) then
                 render.SetBlend(0)
-                self.o.Draw(self)
+                if self.o.Draw ~= TARDIS.DrawOverride then
```
Comment (AmyJeanes) [inline lua/tardis/sh_parts.lua]: Is this check needed? This will be called potentially 30+ times per frame

Comment (AmyJeanes) [reply lua/tardis/sh_parts.lua]: Removed for now after discussion

**Review by AmyJeanes — APPROVED** [review]

**Pull request merged**